File: rpyyarv/scripts/microbench.py
#!/usr/bin/env python3
"""Run benchmark/*.rb on every available engine and compare best-of-N wall times.

Every benchmark prints a checksum, so a run is only reported when all engines
agree on stdout; disagreement is flagged as MISMATCH and exits nonzero.
"""

import argparse
import os
import platform
import subprocess
import sys
import time

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
TOP = os.path.dirname(ROOT)
BUILD = os.environ.get("RPYYARV_BUILD", os.path.join(TOP, "build"))

# Shared-library lookup for the CRuby build tree the engines link against.
LIBVAR = "DYLD_LIBRARY_PATH" if platform.system() == "Darwin" else "LD_LIBRARY_PATH"

ENGINES = [
    ("cruby", [os.path.join(BUILD, "ruby"), "--disable-gems"]),
    ("cruby+yjit", [os.path.join(BUILD, "ruby"), "--yjit", "--disable-gems"]),
    ("cruby+zjit", [os.path.join(BUILD, "ruby"), "--zjit", "--disable-gems"]),
    ("rpyyarv", [os.path.join(ROOT, "rpyyarv")]),
    ("rpyyarv-jit", [os.path.join(ROOT, "rpyyarv-jit")]),
]

CRUBY_ENGINES = ("cruby", "cruby+yjit", "cruby+zjit")


def env_for_run():
    env = dict(os.environ)
    env[LIBVAR] = BUILD + os.pathsep + env.get(LIBVAR, "")
    return env


def run_once(argv, script, env):
    start = time.perf_counter()
    proc = subprocess.run(argv + [script], stdout=subprocess.PIPE,
                          stderr=subprocess.PIPE, env=env)
    elapsed = time.perf_counter() - start
    if proc.returncode != 0:
        # The boot path always warns about unloaded default gems; drop that noise.
        err = "\n".join(l for l in proc.stderr.decode("utf-8", "replace").splitlines()
                        if "not loaded" not in l).strip()
        return None, elapsed, "exit %d: %s" % (proc.returncode, err[-300:])
    return proc.stdout.decode("utf-8", "replace").strip(), elapsed, None


def flags_supported(argv, env):
    """A JIT flag the build lacks makes ruby exit nonzero before running anything."""
    proc = subprocess.run(argv + ["-e", ""], stdout=subprocess.DEVNULL,
                          stderr=subprocess.DEVNULL, env=env)
    return proc.returncode == 0


def fmt(value):
    return "%.3f" % value if value is not None else "-"


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--reps", type=int, default=3, help="repetitions per engine")
    ap.add_argument("--filter", default="", help="only benchmarks whose name contains this")
    args = ap.parse_args()

    bench_dir = os.path.join(ROOT, "benchmark")
    scripts = sorted(f for f in os.listdir(bench_dir) if f.endswith(".rb"))
    scripts = [f for f in scripts if args.filter in f]
    if not scripts:
        print("no benchmarks matched", file=sys.stderr)
        return 1

    env = env_for_run()
    engines = []
    for name, argv in ENGINES:
        if not (os.path.exists(argv[0]) and os.access(argv[0], os.X_OK)):
            print("note: skipping %s (%s not found)" % (name, argv[0]))
        elif not flags_supported(argv, env):
            print("note: skipping %s (flags unsupported by this build)" % name)
        else:
            engines.append((name, argv))
    if not engines:
        print("no engines available", file=sys.stderr)
        return 1

    rows = []
    failed = False

    for script in scripts:
        path = os.path.join(bench_dir, script)
        name = script[:-3]
        times, outputs, notes = {}, {}, []
        for ename, argv in engines:
            best = None
            for _ in range(args.reps):
                out, elapsed, err = run_once(argv, path, env)
                if err is not None:
                    notes.append("%s CRASH (%s)" % (ename, err))
                    best = None
                    break
                outputs.setdefault(ename, out)
                if outputs[ename] != out:
                    notes.append("%s UNSTABLE OUTPUT across reps" % ename)
                best = elapsed if best is None else min(best, elapsed)
            times[ename] = best
            if best is None:
                failed = True

        distinct = set(outputs.values())
        if len(distinct) > 1:
            failed = True
            notes.append("MISMATCH: " + "; ".join(
                "%s=%r" % (e, o) for e, o in sorted(outputs.items())))
        checksum = sorted(distinct)[0] if len(distinct) == 1 else "?"
        rows.append((name, checksum, times, notes))
        print("ran %-14s checksum=%s%s" % (name, checksum,
                                           "  [" + " | ".join(notes) + "]" if notes else ""))

    print()
    headers = ["benchmark"] + [e for e, _ in engines] + ["jit/cruby*"]
    table = []
    for name, checksum, times, _ in rows:
        cells = [name] + [fmt(times.get(e)) for e, _ in engines]
        # Ratio is against the fastest CRuby variant, whichever engine that is.
        best_c = min([times[e] for e in CRUBY_ENGINES if times.get(e)] or [None])
        j = times.get("rpyyarv-jit")
        cells.append("%.2fx" % (j / best_c) if best_c and j else "-")
        table.append(cells)

    widths = [max(len(r[i]) for r in [headers] + table) for i in range(len(headers))]
    line = lambda r: "  ".join(
        c.ljust(widths[i]) if i == 0 else c.rjust(widths[i]) for i, c in enumerate(r))
    print(line(headers))
    print("  ".join("-" * w for w in widths))
    for r in table:
        print(line(r))
    print("\nbest of %d reps, seconds; cruby* = fastest CRuby variant per row" % args.reps)

    if failed:
        print("\nFAILED: see notes above", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: rpyyarv/scripts/microbench.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# Run benchmark/*.rb on every available engine and compare best-of-N wall times.
#
# Every benchmark prints a checksum, so a run is only reported when all engines
# agree on stdout; disagreement is flagged as MISMATCH and exits nonzero.

require "open3"

HERE = File.dirname(File.expand_path(__FILE__))
ROOT = File.dirname(HERE)
TOP = File.dirname(ROOT)
BUILD = ENV.fetch("RPYYARV_BUILD", File.join(TOP, "build"))

# Shared-library lookup for the CRuby build tree the engines link against.
LIBVAR = RUBY_PLATFORM.include?("darwin") ? "DYLD_LIBRARY_PATH" : "LD_LIBRARY_PATH"

ENGINES = [
  ["cruby",       [File.join(BUILD, "ruby"), "--disable-gems"]],
  ["cruby+yjit",  [File.join(BUILD, "ruby"), "--yjit", "--disable-gems"]],
  ["cruby+zjit",  [File.join(BUILD, "ruby"), "--zjit", "--disable-gems"]],
  ["rpyyarv",     [File.join(ROOT, "rpyyarv")]],
  ["rpyyarv-jit", [File.join(ROOT, "rpyyarv-jit")]]
].freeze

CRUBY_ENGINES = ["cruby", "cruby+yjit", "cruby+zjit"].freeze

def env_for_run
  { LIBVAR => BUILD + File::PATH_SEPARATOR + (ENV[LIBVAR] || "") }
end

def run_once(argv, script, env)
  start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  out, err, status = Open3.capture3(env, *argv, script)
  elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
  unless status.success?
    # The boot path always warns about unloaded default gems; drop that noise.
    msg = err.lines.map(&:chomp).reject { |l| l.include?("not loaded") }.join("\n").strip
    return [nil, elapsed, format("exit %d: %s", status.exitstatus.to_i, msg[-300..] || msg)]
  end
  [out.strip, elapsed, nil]
end

# A JIT flag the build lacks makes ruby exit nonzero before running anything.
def flags_supported?(argv, env)
  system(env, *argv, "-e", "", out: File::NULL, err: File::NULL)
end

def fmt(value)
  value ? format("%.3f", value) : "-"
end

def main(argv)
  reps = 3
  filter = ""
  until argv.empty?
    arg = argv.shift
    case arg
    when /\A--reps=(.*)\z/ then reps = Regexp.last_match(1).to_i
    when "--reps" then reps = argv.shift.to_i
    when /\A--filter=(.*)\z/ then filter = Regexp.last_match(1)
    when "--filter" then filter = argv.shift.to_s
    when "-h", "--help"
      puts "usage: microbench.rb [--reps N] [--filter SUBSTRING]"
      return 0
    else
      warn "unrecognized argument: #{arg}"
      return 2
    end
  end

  bench_dir = File.join(ROOT, "benchmark")
  scripts = Dir.children(bench_dir).select { |f| f.end_with?(".rb") }.sort
  scripts = scripts.select { |f| f.include?(filter) }
  if scripts.empty?
    warn "no benchmarks matched"
    return 1
  end

  env = env_for_run
  engines = []
  ENGINES.each do |name, eargv|
    if !File.executable?(eargv[0])
      puts format("note: skipping %s (%s not found)", name, eargv[0])
    elsif !flags_supported?(eargv, env)
      puts format("note: skipping %s (flags unsupported by this build)", name)
    else
      engines << [name, eargv]
    end
  end
  if engines.empty?
    warn "no engines available"
    return 1
  end

  rows = []
  failed = false

  scripts.each do |script|
    path = File.join(bench_dir, script)
    name = script[0..-4]
    times = {}
    outputs = {}
    notes = []
    engines.each do |ename, eargv|
      best = nil
      reps.times do
        out, elapsed, err = run_once(eargv, path, env)
        unless err.nil?
          notes << format("%s CRASH (%s)", ename, err)
          best = nil
          break
        end
        outputs[ename] = out unless outputs.key?(ename)
        notes << format("%s UNSTABLE OUTPUT across reps", ename) if outputs[ename] != out
        best = best.nil? ? elapsed : [best, elapsed].min
      end
      times[ename] = best
      failed = true if best.nil?
    end

    distinct = outputs.values.uniq
    if distinct.size > 1
      failed = true
      notes << "MISMATCH: " + outputs.sort.map { |e, o| format("%s=%s", e, o.inspect) }.join("; ")
    end
    checksum = distinct.size == 1 ? distinct.sort[0] : "?"
    rows << [name, checksum, times, notes]
    puts format("ran %-14s checksum=%s%s", name, checksum,
                notes.empty? ? "" : "  [" + notes.join(" | ") + "]")
  end

  puts
  headers = ["benchmark"] + engines.map { |e, _| e } + ["speedup"]
  table = []
  rows.each do |name, _checksum, times, _notes|
    cells = [name] + engines.map { |e, _| fmt(times[e]) }
    # Speedup vs the fastest CRuby variant: >1 means rpyyarv-jit is faster.
    best_c = CRUBY_ENGINES.map { |e| times[e] }.compact.min
    j = times["rpyyarv-jit"]
    cells << (best_c && j && j > 0 ? format("%.2fx", best_c / j) : "-")
    table << cells
  end

  widths = (0...headers.size).map { |i| ([headers] + table).map { |r| r[i].length }.max }
  line = lambda do |r|
    r.each_with_index.map { |c, i| i.zero? ? c.ljust(widths[i]) : c.rjust(widths[i]) }.join("  ")
  end
  puts line.call(headers)
  puts widths.map { |w| "-" * w }.join("  ")
  table.each { |r| puts line.call(r) }
  puts format("\nbest of %d reps, seconds; speedup = fastest CRuby / rpyyarv-jit", reps)

  if failed
    warn "\nFAILED: see notes above"
    return 1
  end
  0
end

exit(main(ARGV)) if __FILE__ == $PROGRAM_NAME
